adr/0025-composition.md

```diff
@@ -0,0 +1,282 @@
+
+Neither of these criteria are met in the context of Zarf components and packages. Here is a practical example from
+[`defenseunicorns/uds-package-mattermost`](https://github.com/defenseunicorns/uds-package-mattermost/blob/5e02c2ceb7b0e097b7e6eb356b19eaff4c913613/zarf.yaml):
+
+1. The `mattermost-(upstream|registry1)` component flavors depend on a `common` "skeleton package".
+   The `common` package cannot be deployed independantly, which violates the "self-contained" principle.
+2. The `mattermost-plugins` component is not "stateless". It must be declared first because it
+   builds a container image during `onCreate` that is expected by the other components.
+3. `mattermost-plugins` is not "self-contained" because, in order to use it,
+   you must override Helm chart values declared by other components.
+
+```yaml
+kind: ZarfPackageConfig
+metadata:
+  name: mattermost
+  description: "UDS Mattermost Package"
+  version: "9.9.0-uds.0"
+
+variables:
+  - name: SUBDOMAIN
+    description: "Subdomain for Mattermost"
+    default: "chat"
+  - name: DOMAIN
+    default: "uds.dev"
+  - name: ACCESS_KEY
+    description: "Access Key for S3 compatible storage"
+  - name: SECRET_KEY
+    description: "Secret Key for S3 compatible storage"
+  - name: DB_PASSWORD
+    description: "Database Password for Mattermost"
+
+components:
+  - name: mattermost-plugins
+    required: true
+    images:
+      - uds-package-mattermost/mattermost-extra-plugins:latest
+    actions:
+      onCreate:
+        before:
+          - dir: plugins
+            cmd: |
+              docker build . -t uds-package-mattermost/mattermost-extra-plugins:latest
+
+  - name: mattermost
+    required: true
+    description: "Deploy Mattermost"
+    import:
+      path: common
+    only:
+      flavor: upstream
+    charts:
+      - name: mattermost-enterprise-edition
+        valuesFiles:
+          - values/upstream-values.yaml
+    images:
+      - appropriate/curl:latest
+      - mattermost/mattermost-enterprise-edition:9.9.0
+
+  - name: mattermost
+    required: true
+    description: "Deploy Mattermost"
+    import:
+      path: common
+    only:
+      flavor: registry1
+      cluster:
+        architecture: amd64
+    charts:
+      - name: mattermost-enterprise-edition
+        valuesFiles:
+          - values/registry1-values.yaml
+    images:
+      - registry1.dso.mil/ironbank/redhat/ubi/ubi9-minimal:9.4
+      - registry1.dso.mil/ironbank/opensource/mattermost/mattermost:9.9.0
+```
+
+### Proposed Solutions
+
+1. `components[].extends` (`string`): delcares this component as an extension (overlay) of another component.
+   Similar to `flavor`, the resulting component is considered the "deployable unit" and cannot be deployed alongside the component it extends.
+2. `components[].requires` (`[]string`): similar to `extends`, declares this component as an extension (overlay) of another component.
```

Comment:
Zarf has discussed boing from `required` -> `optional` but this may conflict with `required` semantics currently.

Reply:
It does conflict, I think any components that use `components[].requires` to declare dependencies on other named components within the package MUST be considered optional. Same with `components[].extends`.

Reply:
How far are we planning on taking this? This basically means that we need to implement a DAG for the components to make sure that we don't end up with cyclical references.

Reply:
@phillebaba is that bad? What prevents cyclical path `imports` right now?

Reply:
Nothing, but that does not mean that we should add to the existing complexity. My feeling is that we are trying to  partially implement Kustomize features in Zarf. Before we do that we should probably look at Kustomize and the challenges it has faced. Mixing declarative YAML with imperative functions is difficult and tends to become messy in complex situations. Which is why Helm is so popular even with all of its flaws. I am no fan of mixing Go templating with YAML but there is a reason people prefer that over Kustomize.

Reply:
https://github.com/defenseunicorns/zarf/blob/014d1d22544c1b85f2444b9cce7584eac6a70ab2/src/pkg/packager/composer/list.go#L166-L174